kv/kvnemesis: TestKVNemesisSingleNode_ReproposalChaos failed

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11116431?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11116431?buildTab=artifacts#/) on master @ [f295bd861a3a427652b19c2254d2401ebb4a3c8e](https://github.com/cockroachdb/cockroach/commits/f295bd861a3a427652b19c2254d2401ebb4a3c8e):


```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode_ReproposalChaos4009697673
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 9086399838791617203
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3343639818
    kvnemesis.go:168: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.Scan(ctx, tk(2337864312806892353), tk(18385962923748999403), 0) // failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690710151.114247405,1->1690710151.113480788,0
          {
            b := &kv.Batch{}
            b.Del(tk(5292134198012247393) /* @s29 */) // omitted
            txn.Run(ctx, b) // omitted
          }
          {
            b := &kv.Batch{}
            b.Del(tk(13931089840599312697) /* @s30 */) // omitted
            b.Put(tk(15374219157100306618), sv(31)) // omitted
            txn.Run(ctx, b) // omitted
          }
          b := &kv.Batch{}
          txn.CommitInBatch(ctx, b) // omitted
          return nil
        }) // failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690710151.114247405,1->1690710151.113480788,0: failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690710151.114247405,1->1690710151.113480788,0
    kvnemesis.go:168: error applying x.Scan(ctx, tk(2337864312806892353), tk(18385962923748999403), 0) // failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690710151.114247405,1->1690710151.113480788,0: failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690710151.114247405,1->1690710151.113480788,0
    kvnemesis.go:203: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3343639818/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3343639818/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3343639818/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis3343639818/kvs-scan.txt
    kvnemesis_test.go:310: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:310
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:247
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisSingleNode_ReproposalChaos
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode_ReproposalChaos4009697673
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (70.38s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30226

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11152829?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11152829?buildTab=artifacts#/) on master @ [8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6](https://github.com/cockroachdb/cockroach/commits/8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6):


```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode_ReproposalChaos1978623890
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 5371465003841823008
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1609272606
    kvnemesis.go:168: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.ReadCommitted)
          {
            b := &kv.Batch{}
            txn.Run(ctx, b) // <nil>
          }
          txn.ReverseScanForUpdate(ctx, tk(77778482246669777), tk(5588773285581782348), 0) // @1690970557.592814464,1 <nil>
          txn.Scan(ctx, tk(6041570819213447631), tk(15977815651010275910), 0) // failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690970557.592814464,2->1690970557.565970637,2
          b := &kv.Batch{}
          txn.CommitInBatch(ctx, b) // omitted
          return nil
        }) // failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690970557.592814464,2->1690970557.565970637,2: failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690970557.592814464,2->1690970557.565970637,2
    kvnemesis.go:168: error applying x.Scan(ctx, tk(6041570819213447631), tk(15977815651010275910), 0) // failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690970557.592814464,2->1690970557.565970637,2: failed indeterminate commit recovery: programming error: timestamp change by implicitly committed transaction: 1690970557.592814464,2->1690970557.565970637,2
    kvnemesis.go:203: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1609272606/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1609272606/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1609272606/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis1609272606/kvs-scan.txt
    kvnemesis_test.go:310: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:310
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:247
        	Error:      	Should be zero, but was 2
        	Test:       	TestKVNemesisSingleNode_ReproposalChaos
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode_ReproposalChaos1978623890
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (67.88s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11342220?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11342220?buildTab=artifacts#/) on master @ [e0de1f12d4f496fc7e4050a33e96f8c635c7a27c](https://github.com/cockroachdb/cockroach/commits/e0de1f12d4f496fc7e4050a33e96f8c635c7a27c):

Fatal error:

```
panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/Batch. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 3021308 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc002478000?, {0x69b39e8, 0xc01169da70})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x4b59b00, 0xc011f013b0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x69b39e8?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x12e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:225
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaDecoder).createTracingSpans(0xc00a739658, {0x69b39e8, 0xc00e5c2ed0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_decoder.go:154 +0x4f0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaDecoder).DecodeAndBind(0xc00a739200?, {0x69b39e8, 0xc00e5c2ed0}, {0xc0119648d0?, 0x5003755?, 0x10?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_decoder.go:63 +0x66
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).Decode(...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:142
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:878 +0x73a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:728 +0x1b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc0186ff500, 0xc00cb78f00?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:689 +0x14b
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc005003500, {0x69b39e8, 0xc01169da70}, {0x6999690, 0xc0186ff500}, 0xc0186fea80)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:418 +0x195
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x69b39e8?, 0xc01169da70?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisSingleNode_ReproposalChaos
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode_ReproposalChaos900707488
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:283: seed: 5382072640056474552
    kvnemesis_test.go:197: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2911087260
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11519427?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11519427?buildTab=artifacts#/) on master @ [a900aa218054812b782c0b3d130b25296c0d14e3](https://github.com/cockroachdb/cockroach/commits/a900aa218054812b782c0b3d130b25296c0d14e3):


```
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:787
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).isKeyLockedByConflictingTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1812
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).add
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1237
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).getOne
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:828
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).get
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:585
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.mvccGetWithValueHeader
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:1218
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.MVCCGetWithValueHeader
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:1143
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.MVCCGet
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:1110
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.Get
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_get.go:36
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_evaluate.go:484
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_evaluate.go:303
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatchWithServersideRefreshes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_read.go:458
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_read.go:125
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:510
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:185
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:193
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:202
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1336
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1467
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:701
        	            	  | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:812
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169
        	            	Wraps: (2) panic applying step 
        	            	  | db0.GetForUpdateSkipLocked(ctx, tk(16811042229377009131)): only transactional requests can acquire exclusive locks
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (3.92s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvnemesis.TestKVNemesisSingleNode_ReproposalChaos [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11519428?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11519428?buildTab=artifacts#/) on master @ [a900aa218054812b782c0b3d130b25296c0d14e3](https://github.com/cockroachdb/cockroach/commits/a900aa218054812b782c0b3d130b25296c0d14e3):


```
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:787
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).isKeyLockedByConflictingTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1812
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).add
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1237
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).getOne
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:828
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).scan
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:659
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.mvccScanToBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:3874
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.mvccScanToKvs
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:3927
        	            	  | github.com/cockroachdb/cockroach/pkg/storage.MVCCScan
        	            	  | 	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:4150
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.ReverseScan
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_reverse_scan.go:82
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_evaluate.go:484
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_evaluate.go:303
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatchWithServersideRefreshes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_read.go:458
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_read.go:125
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:510
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:185
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:193
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:202
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1336
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1467
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:701
        	            	  | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.bindUnaryServerInterceptorToHandler.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:812
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169
        	            	Wraps: (2) panic applying step 
        	            	  | db0.ReverseScanForUpdateSkipLocked(ctx, tk(11866519064140098950), tk(14356790331432790134), 0): only transactional requests can acquire exclusive locks
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    panic.go:522: -- test log scope end --
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (16.74s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode_ReproposalChaos.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/concurrency/lock_table.go
+++ b/pkg/kv/kvserver/concurrency/lock_table.go
@@ -801,19 +801,17 @@ func (g *lockTableGuardImpl) curLockMode() lock.Mode {
 
 // makeLockMode constructs and returns a lock mode.
 func makeLockMode(str lock.Strength, txn *roachpb.Transaction, ts hlc.Timestamp) lock.Mode {
+	iso := isolation.Serializable
+	if txn != nil {
+		iso = txn.IsoLevel
+	}
 	switch str {
 	case lock.None:
-		iso := isolation.Serializable
-		if txn != nil {
-			iso = txn.IsoLevel
-		}
 		return lock.MakeModeNone(ts, iso)
 	case lock.Shared:
-		assert(txn != nil, "only transactional requests can acquire shared locks")
 		return lock.MakeModeShared()
 	case lock.Exclusive:
-		assert(txn != nil, "only transactional requests can acquire exclusive locks")
-		return lock.MakeModeExclusive(ts, txn.IsoLevel)
+		return lock.MakeModeExclusive(ts, iso)
 	case lock.Intent:
 		return lock.MakeModeIntent(ts)
 	default:

